File: mobile/lib/services/chat_service.dart
import 'dart:math';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:flutter/foundation.dart';
import '../models/chat_message_model.dart';
import '../models/chat_user_profile.dart';
import '../models/conversation_model.dart';
import 'firebase_service.dart';

const String _allowedChars = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789';

class ChatService {
  static FirebaseFirestore get _firestore => FirebaseService.firestore;

  /// Generates a random 5-character uppercase alphanumeric ID avoiding ambiguous characters.
  static String generateRandomShortId() {
    final rand = Random.secure();
    return List.generate(
      5,
      (_) => _allowedChars[rand.nextInt(_allowedChars.length)],
    ).join();
  }

  /// Ensures the user has a permanent unique 5-character shortUserId.
  /// Strictly reproduces planner/src/lib/firebase/chatService.ts ensureUserChatId.
  static Future<String> ensureUserChatId(String uid, [String? displayName]) async {
    if (uid.isEmpty) throw ArgumentError('User UID is required');

    final cleanName = displayName?.trim() ?? '';
    final userRef = _firestore.collection('users').doc(uid);
    final userSnap = await userRef.get();

    if (userSnap.exists && userSnap.data()?['shortUserId'] != null) {
      final existingShortId = userSnap.data()!['shortUserId'] as String;
      if (cleanName.isNotEmpty && userSnap.data()?['displayName'] != cleanName) {
        await userRef.set({
          'displayName': cleanName,
          'updatedAt': FieldValue.serverTimestamp(),
        }, SetOptions(merge: true));

        final lookupRef = _firestore.collection('userIds').doc(existingShortId);
        await lookupRef.set({
          'displayName': cleanName,
        }, SetOptions(merge: true));
      }
      return existingShortId;
    }

    // Generate unique Chat ID using Firestore transaction
    int attempts = 0;
    while (attempts < 10) {
      attempts++;
      final candidateId = generateRandomShortId();
      final userIdLookupRef = _firestore.collection('userIds').doc(candidateId);

      try {
        final assignedId = await _firestore.runTransaction<String>((transaction) async {
          final lookupSnap = await transaction.get(userIdLookupRef);
          if (lookupSnap.exists) {
            throw Exception('COLLISION');
          }

          transaction.set(
            userRef,
            {
              'shortUserId': candidateId,
              'displayName': cleanName,
              'createdAt': FieldValue.serverTimestamp(),
              'updatedAt': FieldValue.serverTimestamp(),
            },
            SetOptions(merge: true),
          );

          transaction.set(
            userIdLookupRef,
            {
              'uid': uid,
              'displayName': cleanName,
              'createdAt': FieldValue.serverTimestamp(),
            },
          );

          return candidateId;
        });

        return assignedId;
      } catch (err) {
        if (err.toString().contains('COLLISION')) {
          continue;
        }
        debugPrint('Error during Chat ID generation transaction: $err');
        rethrow;
      }
    }

    throw Exception('Failed to generate unique Chat ID after maximum attempts');
  }

  /// Finds a user's Firebase UID and info by their 5-character Chat ID.
  static Future<Map<String, dynamic>?> lookupUserByChatId(String shortUserId) async {
    final sanitizedId = shortUserId.trim().toUpperCase();
    if (sanitizedId.length < 4) return null;

    final lookupRef = _firestore.collection('userIds').doc(sanitizedId);
    final lookupSnap = await lookupRef.get();

    if (lookupSnap.exists) {
      final data = lookupSnap.data();
      return {
        'uid': data?['uid'] as String?,
        'displayName': data?['displayName'] as String?,
      };
    }
    return null;
  }

  /// Gets short Chat ID for a given user UID.
  static Future<String?> getUserShortId(String uid) async {
    final userRef = _firestore.collection('users').doc(uid);
    final userSnap = await userRef.get();
    if (userSnap.exists) {
      return userSnap.data()?['shortUserId'] as String?;
    }
    return null;
  }

  /// Gets full UserProfile (shortUserId + displayName) for a given user UID.
  static Future<ChatUserProfile?> getUserProfile(String uid) async {
    final userRef = _firestore.collection('users').doc(uid);
    final userSnap = await userRef.get();
    if (userSnap.exists) {
      final data = userSnap.data() ?? {};
      return ChatUserProfile(
        uid: uid,
        shortUserId: (data['shortUserId'] as String?) ?? '',
        displayName: (data['displayName'] as String?) ?? '',
      );
    }
    return null;
  }

  /// Deterministically creates or retrieves a 1-to-1 conversation between two users.
  static Future<String> getOrCreateConversation(String currentUid, String targetUid) async {
    if (currentUid == targetUid) {
      throw ArgumentError('Cannot message yourself');
    }

    final sortedUids = [currentUid, targetUid]..sort();
    final conversationId = '${sortedUids[0]}_${sortedUids[1]}';
    final convRef = _firestore.collection('conversations').doc(conversationId);

    // Fetch profiles for both participants to embed in the conversation doc
    final profiles = <String, Map<String, String>>{};
    try {
      final currentProf = await getUserProfile(currentUid);
      if (currentProf != null) {
        profiles[currentUid] = {
          'shortUserId': currentProf.shortUserId,
          'displayName': currentProf.displayName,
        };
      }
    } catch (_) {}

    try {
      final targetProf = await getUserProfile(targetUid);
      if (targetProf != null) {
        profiles[targetUid] = {
          'shortUserId': targetProf.shortUserId,
          'displayName': targetProf.displayName,
        };
      }
    } catch (_) {}

    try {
      final convSnap = await convRef.get();
      if (!convSnap.exists) {
        await convRef.set({
          'participants': sortedUids,
          'participantKey': conversationId,
          'participantProfiles': profiles,
          'lastMessage': '',
          'lastMessageAt': FieldValue.serverTimestamp(),
          'lastSenderId': '',
          'createdAt': FieldValue.serverTimestamp(),
        });
      } else if (profiles.isNotEmpty) {
        await convRef.set({
          'participantProfiles': profiles,
        }, SetOptions(merge: true));
      }
    } catch (err) {
      debugPrint('Fallback saving conversation doc: $err');
      await convRef.set({
        'participants': sortedUids,
        'participantKey': conversationId,
        'participantProfiles': profiles,
        'lastMessage': '',
        'lastMessageAt': FieldValue.serverTimestamp(),
        'lastSenderId': '',
        'createdAt': FieldValue.serverTimestamp(),
      }, SetOptions(merge: true));
    }

    return conversationId;
  }

  /// Sends a direct message in a conversation.
  static Future<String> sendMessage({
    required String conversationId,
    required String senderId,
    required String receiverId,
    required String text,
  }) async {
    final cleanText = text.trim();
    if (cleanText.isEmpty) throw ArgumentError('Message text cannot be empty');

    final messagesColRef = _firestore
        .collection('conversations')
        .doc(conversationId)
        .collection('messages');

    final docRef = await messagesColRef.add({
      'senderId': senderId,
      'receiverId': receiverId,
      'text': cleanText,
      'status': 'sent',
      'createdAt': FieldValue.serverTimestamp(),
      'deliveredAt': FieldValue.serverTimestamp(),
      'seenAt': null,
    });

    // Update last message preview and increment recipient's unread count atomically
    final convRef = _firestore.collection('conversations').doc(conversationId);
    await convRef.update({
      'lastMessage': cleanText,
      'lastMessageAt': FieldValue.serverTimestamp(),
      'lastSenderId': senderId,
      'unreadCounts.$receiverId': FieldValue.increment(1),
    });

    return docRef.id;
  }

  /// Marks incoming unread messages in a conversation as seen and resets recipient's unread count to 0.
  static Future<void> markMessagesAsSeen(String conversationId, String currentUid) async {
    if (conversationId.isEmpty || currentUid.isEmpty) return;

    // Reset unread count for current user on conversation document
    final convRef = _firestore.collection('conversations').doc(conversationId);
    await convRef.update({
      'unreadCounts.$currentUid': 0,
    }).catchError((_) {});

    try {
      final messagesColRef = _firestore
          .collection('conversations')
          .doc(conversationId)
          .collection('messages');

      final querySnapshot = await messagesColRef
          .where('receiverId', isEqualTo: currentUid)
          .where('status', isEqualTo: 'sent')
          .get();

      if (querySnapshot.docs.isEmpty) return;

      final batch = _firestore.batch();
      for (final messageDoc in querySnapshot.docs) {
        batch.update(messageDoc.reference, {
          'status': 'seen',
          'seenAt': FieldValue.serverTimestamp(),
        });
      }

      await batch.commit();
    } catch (err) {
      debugPrint('Error marking messages as seen: $err');
    }
  }

  /// Real-time stream for user's conversations list, sorted by lastMessageAt descending.
  static Stream<List<ConversationModel>> getConversationsStream(String currentUid) {
    return _firestore
        .collection('conversations')
        .where('participants', arrayContains: currentUid)
        .snapshots()
        .map((snapshot) {
      final list = snapshot.docs
          .map((doc) => ConversationModel.fromFirestore(doc, currentUid))
          .toList();

      list.sort((a, b) {
        final aTime = a.lastMessageAt ?? DateTime.fromMillisecondsSinceEpoch(0);
        final bTime = b.lastMessageAt ?? DateTime.fromMillisecondsSinceEpoch(0);
        return bTime.compareTo(aTime);
      });

      return list;
    });
  }

  /// Real-time stream for messages within a specific conversation.
  static Stream<List<ChatMessageModel>> getMessagesStream(String conversationId) {
    return _firestore
        .collection('conversations')
        .doc(conversationId)
        .collection('messages')
        .orderBy('createdAt', descending: false)
        .snapshots()
        .map((snapshot) {
      return snapshot.docs.map((doc) => ChatMessageModel.fromFirestore(doc)).toList();
    });
  }

  /// Stream of total unread messages count for badge display.
  static Stream<int> getTotalUnreadCountStream(String currentUid) {
    return _firestore
        .collection('conversations')
        .where('participants', arrayContains: currentUid)
        .snapshots()
        .map((snapshot) {
      int total = 0;
      for (final doc in snapshot.docs) {
        final data = doc.data();
        if (data['unreadCounts'] is Map) {
          final unreads = data['unreadCounts'] as Map;
          final count = unreads[currentUid];
          if (count is num) {
            total += count.toInt();
          }
        }
      }
      return total;
    });
  }
}

File: planner/src/lib/firebase/chatService.ts
import { 
  doc, 
  getDoc, 
  getDocs, 
  setDoc, 
  updateDoc, 
  addDoc, 
  collection, 
  query, 
  where, 
  orderBy, 
  onSnapshot, 
  runTransaction, 
  writeBatch, 
  serverTimestamp,
  increment,
  Timestamp,
  Unsubscribe
} from 'firebase/firestore'
import { db } from './config'

export interface ChatUser {
  uid: string
  shortUserId: string
  createdAt?: any
  updatedAt?: any
}

export interface Conversation {
  id: string
  participants: string[]
  participantKey: string
  participantProfiles?: Record<string, { shortUserId: string; displayName: string }>
  unreadCounts?: Record<string, number>
  lastMessage: string
  lastMessageAt: any
  lastSenderId: string
  createdAt: any
  otherUid: string
  otherUserShortId?: string
}

export interface Message {
  id: string
  senderId: string
  receiverId: string
  text: string
  status: 'sending' | 'sent' | 'seen'
  createdAt: any
  deliveredAt?: any
  seenAt?: any
}

const ALLOWED_CHARS = 'ABCDEFGHJKLMNPQRSTUVWXYZ23456789'

export interface UserProfile {
  uid: string
  shortUserId: string
  displayName?: string
}

/**
 * Generates a random 5-character uppercase alphanumeric ID avoiding ambiguous characters.
 */
export function generateRandomShortId(): string {
  let result = ''
  for (let i = 0; i < 5; i++) {
    const randomIndex = Math.floor(Math.random() * ALLOWED_CHARS.length)
    result += ALLOWED_CHARS.charAt(randomIndex)
  }
  return result
}

/**
 * Ensures the user has a permanent unique 5-character shortUserId.
 */
export async function ensureUserChatId(uid: string, displayName?: string): Promise<string> {
  if (!uid) throw new Error('User UID is required')

  const cleanName = displayName?.trim() || ''
  const userRef = doc(db, 'users', uid)
  const userSnap = await getDoc(userRef)

  if (userSnap.exists() && userSnap.data()?.shortUserId) {
    const existingShortId = userSnap.data().shortUserId
    if (cleanName && userSnap.data()?.displayName !== cleanName) {
      await setDoc(userRef, { displayName: cleanName, updatedAt: serverTimestamp() }, { merge: true })
      const lookupRef = doc(db, 'userIds', existingShortId)
      await setDoc(lookupRef, { displayName: cleanName }, { merge: true })
    }
    return existingShortId
  }

  // Generate unique Chat ID using Firestore transaction
  let attempts = 0
  while (attempts < 10) {
    attempts++
    const candidateId = generateRandomShortId()
    const userIdLookupRef = doc(db, 'userIds', candidateId)

    try {
      const assignedId = await runTransaction(db, async (transaction) => {
        const lookupSnap = await transaction.get(userIdLookupRef)
        if (lookupSnap.exists()) {
          throw new Error('COLLISION')
        }

        transaction.set(userRef, {
          shortUserId: candidateId,
          displayName: cleanName,
          createdAt: serverTimestamp(),
          updatedAt: serverTimestamp()
        }, { merge: true })

        transaction.set(userIdLookupRef, {
          uid: uid,
          displayName: cleanName,
          createdAt: serverTimestamp()
        })

        return candidateId
      })

      return assignedId
    } catch (err: any) {
      if (err.message === 'COLLISION') {
        continue
      }
      console.error('Error during Chat ID generation transaction:', err)
      throw err
    }
  }

  throw new Error('Failed to generate unique Chat ID after maximum attempts')
}

/**
 * Finds a user's Firebase UID by their 5-character Chat ID.
 */
export async function lookupUserByChatId(shortUserId: string): Promise<string | null> {
  const sanitizedId = shortUserId.trim().toUpperCase()
  if (!sanitizedId || sanitizedId.length < 4) return null

  const lookupRef = doc(db, 'userIds', sanitizedId)
  const lookupSnap = await getDoc(lookupRef)

  if (lookupSnap.exists()) {
    return lookupSnap.data()?.uid || null
  }
  return null
}

/**
 * Gets short Chat ID for a given user UID.
 */
export async function getUserShortId(uid: string): Promise<string | null> {
  const userRef = doc(db, 'users', uid)
  const userSnap = await getDoc(userRef)
  if (userSnap.exists()) {
    return userSnap.data()?.shortUserId || null
  }
  return null
}

/**
 * Gets full UserProfile (shortUserId + displayName) for a given user UID.
 */
export async function getUserProfile(uid: string): Promise<UserProfile | null> {
  const userRef = doc(db, 'users', uid)
  const userSnap = await getDoc(userRef)
  if (userSnap.exists()) {
    const data = userSnap.data()
    return {
      uid,
      shortUserId: data?.shortUserId || '',
      displayName: data?.displayName || ''
    }
  }
  return null
}

/**
 * Deterministically creates or retrieves a 1-to-1 conversation between two users.
 */
export async function getOrCreateConversation(currentUid: string, targetUid: string): Promise<string> {
  if (currentUid === targetUid) {
    throw new Error('Cannot message yourself')
  }

  const sortedUids = [currentUid, targetUid].sort()
  const conversationId = `${sortedUids[0]}_${sortedUids[1]}`
  const convRef = doc(db, 'conversations', conversationId)

  // Fetch profiles for both participants to embed in the conversation doc
  const [currentProf, targetProf] = await Promise.all([
    getUserProfile(currentUid).catch(() => null),
    getUserProfile(targetUid).catch(() => null)
  ])

  const profiles: Record<string, { shortUserId: string; displayName: string }> = {}
  if (currentProf) profiles[currentUid] = { shortUserId: currentProf.shortUserId, displayName: currentProf.displayName || '' }
  if (targetProf) profiles[targetUid] = { shortUserId: targetProf.shortUserId, displayName: targetProf.displayName || '' }

  try {
    const convSnap = await getDoc(convRef)
    if (!convSnap.exists()) {
      await setDoc(convRef, {
        participants: sortedUids,
        participantKey: conversationId,
        participantProfiles: profiles,
        lastMessage: '',
        lastMessageAt: serverTimestamp(),
        lastSenderId: '',
        createdAt: serverTimestamp()
      })
    } else if (Object.keys(profiles).length > 0) {
      await setDoc(convRef, { participantProfiles: profiles }, { merge: true })
    }
  } catch (err: any) {
    await setDoc(convRef, {
      participants: sortedUids,
      participantKey: conversationId,
      participantProfiles: profiles,
      lastMessage: '',
      lastMessageAt: serverTimestamp(),
      lastSenderId: '',
      createdAt: serverTimestamp()
    }, { merge: true })
  }

  return conversationId
}

/**
 * Sends a direct message in a conversation.
 */
export async function sendMessage(
  conversationId: string, 
  senderId: string, 
  receiverId: string, 
  text: string
): Promise<string> {
  const cleanText = text.trim()
  if (!cleanText) throw new Error('Message text cannot be empty')

  const messagesColRef = collection(db, 'conversations', conversationId, 'messages')
  
  const docRef = await addDoc(messagesColRef, {
    senderId,
    receiverId,
    text: cleanText,
    status: 'sent',
    createdAt: serverTimestamp(),
    deliveredAt: serverTimestamp(),
    seenAt: null
  })

  // Update last message preview and increment recipient's unread count atomically
  const convRef = doc(db, 'conversations', conversationId)
  await updateDoc(convRef, {
    lastMessage: cleanText,
    lastMessageAt: serverTimestamp(),
    lastSenderId: senderId,
    [`unreadCounts.${receiverId}`]: increment(1)
  })

  return docRef.id
}

/**
 * Marks incoming unread messages in a conversation as seen and resets recipient's unread count to 0.
 */
export async function markMessagesAsSeen(conversationId: string, currentUid: string): Promise<void> {
  if (!conversationId || !currentUid) return

  // Reset unread count for current user on conversation document
  const convRef = doc(db, 'conversations', conversationId)
  await updateDoc(convRef, {
    [`unreadCounts.${currentUid}`]: 0
  }).catch(() => {})

  try {
    const messagesColRef = collection(db, 'conversations', conversationId, 'messages')
    const q = query(
      messagesColRef,
      where('receiverId', '==', currentUid),
      where('status', '==', 'sent')
    )

    const querySnapshot = await getDocs(q)
    if (querySnapshot.empty) return

    const batch = writeBatch(db)
    querySnapshot.docs.forEach((messageDoc: any) => {
      batch.update(messageDoc.ref, {
        status: 'seen',
        seenAt: serverTimestamp()
      })
    })

    await batch.commit()
  } catch (err) {
    console.error('Error marking messages as seen:', err)
  }
}

/**
 * Real-time listener for user's conversations list.
 */
export function subscribeToConversations(
  currentUid: string, 
  onUpdate: (conversations: Conversation[]) => void
): Unsubscribe {
  const convsColRef = collection(db, 'conversations')
  const q = query(
    convsColRef,
    where('participants', 'array-contains', currentUid)
  )

  return onSnapshot(q, async (snapshot) => {
    const rawConvs: Conversation[] = []
    
    for (const docSnap of snapshot.docs) {
      const data = docSnap.data()
      const otherUid = (data.participants as string[]).find(uid => uid !== currentUid) || currentUid
      const profiles = data.participantProfiles || {}
      const targetProfile = profiles[otherUid]
      
      rawConvs.push({
        id: docSnap.id,
        participants: data.participants,
        participantKey: data.participantKey,
        participantProfiles: profiles,
        unreadCounts: data.unreadCounts || {},
        lastMessage: data.lastMessage || '',
        lastMessageAt: data.lastMessageAt,
        lastSenderId: data.lastSenderId || '',
        createdAt: data.createdAt,
        otherUid: otherUid,
        otherUserShortId: targetProfile?.shortUserId || ''
      })
    }

    // Sort client-side by lastMessageAt descending to avoid needing a Firestore composite index
    rawConvs.sort((a, b) => {
      const getMillis = (t: any) => {
        if (!t) return 0
        if (typeof t.toMillis === 'function') return t.toMillis()
        if (t.seconds) return t.seconds * 1000
        return 0
      }
      return getMillis(b.lastMessageAt) - getMillis(a.lastMessageAt)
    })

    onUpdate(rawConvs)
  }, (err) => {
    console.error('Error listening to conversations:', err)
  })
}

/**
 * Real-time listener for messages within a specific conversation.
 */
export function subscribeToMessages(
  conversationId: string, 
  onUpdate: (messages: Message[]) => void
): Unsubscribe {
  const messagesColRef = collection(db, 'conversations', conversationId, 'messages')
  const q = query(messagesColRef, orderBy('createdAt', 'asc'))

  return onSnapshot(q, (snapshot) => {
    const messages: Message[] = snapshot.docs.map(docSnap => {
      const data = docSnap.data()
      return {
        id: docSnap.id,
        senderId: data.senderId,
        receiverId: data.receiverId,
        text: data.text,
        status: data.status || 'sent',
        createdAt: data.createdAt,
        deliveredAt: data.deliveredAt,
        seenAt: data.seenAt
      }
    })
    onUpdate(messages)
  }, (err) => {
    console.error(`Error listening to messages for conversation ${conversationId}:`, err)
  })
}
